api/document/js/Api.ts

```diff
@@ -18,31 +18,60 @@ export function setStatusError(e: Error) {
   setStatus(errMsg, 'editor-status-error');
 }
 
-export default class Api {
+export interface ApiNode {
+  children: ApiNode[];
+  content: ApiContent[];
+  marker?: string;
+  node_type: string;
+  type_emblem?: string;
+  title?: string;
+  identifier?: string;
+  text?: string;
+}
+
+export interface ApiContent {
+  content_type: string;
+  inlines: ApiContent[];
+  text: string;
+}
+
+interface ApiTypeMap {
+  'json': ApiNode;
+  'akn+xml': string;
+}
+
+interface ApiOptions<T extends keyof ApiTypeMap> {
+  contentType: T;
+  csrfToken: string;
+  url: string;
+}
+
+export default class Api<T extends keyof ApiTypeMap> {
```

Comment:
What do you think about making this type signature less abstract? Right now it's `Api<'json'>` or `Api<'akn+xml'>`; would it be easier to understand as `Api<ApiNode>` and `Api<string>`? It might even make sense to lock the type var in by having `NodeApi` and `StringApi` as subclasses of an abstract `Api<T>`.

Reply:
Yeah, that is what I originally wanted, but I wasn't sure how best to map the type into the `content-type` header that ultimately got passed to the server--I didn't want, for example, someone to be able to make an `Api<ApiNode>` that passed a `content-type` of `akn+xml`--so I looked up the type definition for `document.createElement()` for inspiration.  But yeah, it's not ideal...  I think the solution that combines the best of readability + flexibility might be your latter solution there.